### ISSUE TITLE
Add stylegan loss

### ISSUE THREAD

**Issue by borisdayma** [patil-suraj/vit-vqgan#8]
The model is ready but unused.

It requires adding another model and another opt_state to train state.
For simplicity I'd suggest probably using same params as for the main model except probably for the learning rate.

**Comment by pcuenca** [comment]
I suppose we also need to implement a discriminator loss.
* The `stylegan2` implementation uses this: ~~https://github.com/NVlabs/stylegan2/blob/master/training/loss.py#L148~~ https://github.com/NVlabs/stylegan2/blob/master/training/loss.py#L52
* The JAX/haiku implementation from stylegan2.jax uses this: https://github.com/dsuess/stylegan2.jax/blob/master/stylegan2/train.py#L96
* This minimalistic example seems to use Wasserstein loss: https://nn.labml.ai/gan/stylegan/experiment.html
* Phil Wang uses hinge_loss or dual_contrastive_loss: https://github.com/lucidrains/stylegan2-pytorch/blob/master/stylegan2_pytorch/stylegan2_pytorch.py#L1010  

I would go with `hinge_loss` to begin with.

**Comment by borisdayma** [comment]
We can try any method, Suraj already added a model but not the loss yet.

The authors of the ViT-VQGAN recommended to follow implementation from [NVlabs/stylegan2](https://github.com/NVlabs/stylegan2) as close as possible and that they used a StyleGAN discriminator and gradient penalty loss term, training generator and discriminator all together from scratch.

**Comment by pcuenca** [comment]
Yes, they mention in the paper that they follow that architecture but without giving any details, it's good they confirmed it later. I'll see if I can replicate that loss then.

**Comment by patil-suraj** [comment]
I didn't see this thread but have started working on it over the weekend. See #17. Will post more details once I test it out.
